sources/host-ctr/cmd/host-ctr/registry.go

```diff
@@ -25,3 +44,94 @@ func NewRegistryConfig(registryConfigFile string) (*RegistryConfig, error) {
 	config := RegistryConfig{}
 	return &config, toml.Unmarshal(raw, &config)
 }
+
+// registryHosts returns the registry hosts to be used by the resolver.
+// Heavily borrowed from containerd CRI plugin's implementation.
+// See https://github.com/containerd/containerd/blob/1407cab509ff0d96baa4f0eb6ff9980270e6e620/pkg/cri/server/image_pull.go#L332-L405
+// authorizerOverride lets the caller override the generated authorizer with a custom authorizer
+// FIXME Replace this once there's a public containerd client interface that supports registry mirrors
+func registryHosts(registryConfig *RegistryConfig, authorizerOverride *docker.Authorizer) docker.RegistryHosts {
+	return func(host string) ([]docker.RegistryHost, error) {
+		var (
+			registries []docker.RegistryHost
+			endpoints  []string
+			authConfig runtime.AuthConfig
+			authorizer docker.Authorizer
+		)
+		// Set up endpoints for the registry
+		if _, ok := registryConfig.Mirrors[host]; ok {
+			endpoints = registryConfig.Mirrors[host].Endpoints
+		} else {
+			endpoints = registryConfig.Mirrors["*"].Endpoints
+		}
+		defaultHost, err := docker.DefaultHost(host)
+		if err != nil {
+			return nil, errors.Wrap(err, "get default host")
+		}
+		endpoints = append(endpoints, defaultHost)
+
+		// Set up auth for pulling from registry
+		var authOpts []docker.AuthorizerOpt
+		if _, ok := registryConfig.Credentials[defaultHost]; ok {
+			// Convert registry credentials config to runtime auth config, so it can be parsed by `ParseAuth`
+			authConfig.Username = registryConfig.Credentials[defaultHost].Username
+			authConfig.Password = registryConfig.Credentials[defaultHost].Password
+			authConfig.Auth = registryConfig.Credentials[defaultHost].Auth
+			authConfig.IdentityToken = registryConfig.Credentials[defaultHost].IdentityToken
+			authOpts = append(authOpts, docker.WithAuthClient(&http.Client{
+				Transport: newTransport(),
+			}))
+			authOpts = append(authOpts, docker.WithAuthCreds(func(host string) (string, string, error) {
+				return server.ParseAuth(&authConfig, host)
+			}))
+		}
+		authorizer = docker.NewDockerAuthorizer(authOpts...)
```

Comment:
Is it OK to share an instance of `authorizer` like this?  The CRI implementation builds one per registry endpoint.

I assume so because the previous implementation did that when passed an authorizer override, although I'm not sure we'd ever end up with more than one in the case where it was used (ecr-public).

It's a bit awkward to do the work to construct the authorizer and then just throw it away if we were given an override.